hazelcast-sql/src/main/java/com/hazelcast/jet/sql/impl/PlanExecutor.java

```diff
@@ -517,64 +518,17 @@ SqlResult execute(SelectPlan plan, QueryId queryId, List<Object> arguments, long
-            // strategy, we should abort the process and clear list of already populated partitions so that partition
-            // pruning doesn't get activated at all for this query.
-            if (strategy != null
-                    && !(strategy instanceof DefaultPartitioningStrategy)
-                    && !(strategy instanceof AttributePartitioningStrategy)) {
-                allVariantsValid = false;
-                break;
-            }
-
-            // ordering of attributes matters for partitioning (1,2) produces different partition than (2,1).
-            final List<String> orderedKeyAttributes = new ArrayList<>();
-            if (strategy instanceof AttributePartitioningStrategy) {
-                final var attributeStrategy = (AttributePartitioningStrategy) strategy;
-                orderedKeyAttributes.addAll(asList(attributeStrategy.getPartitioningAttributes()));
-            } else {
-                orderedKeyAttributes.add(DEFAULT_UNIQUE_KEY);
-            }
-
-            for (final Map<String, Expression<?>> perMapCandidate : perMapCandidates) {
-                Object[] partitionKeyComponents = new Object[orderedKeyAttributes.size()];
-                for (int i = 0; i < orderedKeyAttributes.size(); i++) {
-                    final String attribute = orderedKeyAttributes.get(i);
-                    if (!perMapCandidate.containsKey(attribute)) {
-                        // Shouldn't happen, defensive check in case Opt logic breaks and produces variants
-                        // that do not contain all the required partitioning attributes.
-                        throw new HazelcastException("Partition Pruning candidate"
-                                + " does not contain mandatory attribute: " + attribute);
-                    }
-
-                    partitionKeyComponents[i] = perMapCandidate.get(attribute).eval(null, evalContext);
-                }
+        if (!plan.getPartitionStrategyCandidates().isEmpty()) {
+            var optResult = tryUsePrunability(plan, evalContext);
+            final Set<Integer> partitions = Objects.requireNonNull(optResult.f0());
+            final boolean allVariantsValid = Boolean.TRUE.equals(optResult.f1());
 
-                final Partition partition = hazelcastInstance.getPartitionService().getPartition(
-                        PartitioningStrategyUtil.constructAttributeBasedKey(partitionKeyComponents)
-                );
-                if (partition == null) {
-                    // Can happen if the cluster is mid-repartitioning/migration, in this case we revert to
-                    // non-pruning logic. Alternative scenario is if the produced partitioning key somehow invalid.
-                    allVariantsValid = false;
-                    break;
+            if (!partitions.isEmpty() && allVariantsValid) {
+                if (plan.requiredRootPartitionId() != null) {
```

Comment:
you will have conflicts here #25062

Reply:
Ofc I will update this PR after the #25062  merge.